packages/@aws-cdk/toolkit-lib/lib/api/drift/drift.ts

```diff
@@ -0,0 +1,76 @@
+import { format } from 'util';
+import type { DescribeStackDriftDetectionStatusCommandOutput, DescribeStackResourceDriftsCommandOutput } from '@aws-sdk/client-cloudformation';
+import type { IDriftCloudFormationClient, IDriftIoHelper } from './interfaces';
+import { ToolkitError } from '../../toolkit/toolkit-error';
+import { IO } from '../io/private';
+
+/**
+ * Detect drift for a CloudFormation stack and wait for the detection to complete
+ *
+ * @param cfn - a CloudFormation client
+ * @param ioHelper - helper for IO operations
+ * @param stackName - the name of the stack to check for drift
+ * @returns the CloudFormation description of the drift detection results
+ */
+export async function detectStackDrift(
+  cfn: IDriftCloudFormationClient,
+  ioHelper: IDriftIoHelper,
+  stackName: string,
+): Promise<DescribeStackResourceDriftsCommandOutput> {
+  await ioHelper.notify(IO.DEFAULT_TOOLKIT_DEBUG.msg(format('Starting drift detection for stack %s...', stackName)));
+
+  // Start drift detection
+  const driftDetection = await cfn.detectStackDrift({
+    StackName: stackName,
+  });
+
+  // Wait for drift detection to complete
+  const driftStatus = await waitForDriftDetection(cfn, ioHelper, driftDetection.StackDriftDetectionId!);
+
+  if (!driftStatus) {
+    throw new ToolkitError('Drift detection took too long to complete. Aborting');
+  }
+
+  if (driftStatus?.DetectionStatus === 'DETECTION_FAILED') {
+    throw new ToolkitError(
+      `Failed to detect drift: ${driftStatus.DetectionStatusReason || 'No reason provided'}`,
+    );
+  }
+
+  // Get the drift results
+  return cfn.describeStackResourceDrifts({
+    StackName: stackName,
+  });
+}
+
+/**
+ * Wait for a drift detection operation to complete
+ */
+async function waitForDriftDetection(
+  cfn: IDriftCloudFormationClient,
+  ioHelper: IDriftIoHelper,
+  driftDetectionId: string,
+): Promise<DescribeStackDriftDetectionStatusCommandOutput | undefined> {
+  await ioHelper.notify(IO.DEFAULT_TOOLKIT_DEBUG.msg(format('Waiting for drift detection %s to complete...', driftDetectionId)));
+
+  const timeout = 10_000; // if takes longer than 10s, fail
```

Comment:
Is it realistic for a drift detection operation to finish in 10s? This seems short to me.

I think we probably want a max timeout of a minute or more, and periodically print something like `Still checking...`. We could even print

```
Checking resource drift (5/20)
```

Maybe the timeout even needs to be configurable; it's going to be the first change request coming in on this I'd imagine.

Reply:
Good point, I'll see if I can include some way to have it track resource count as they're being checked, or just something to let the user know the command is still churning away